lib/ui/window/platform_message_response_dart_port.cc

```diff
@@ -0,0 +1,72 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/window/platform_message_response_dart_port.h"
+
+#include <utility>
+
+#include "flutter/common/task_runners.h"
+#include "flutter/fml/make_copyable.h"
+#include "flutter/fml/trace_event.h"
+#include "third_party/dart/runtime/include/dart_native_api.h"
+#include "third_party/tonic/converter/dart_converter.h"
+#include "third_party/tonic/dart_state.h"
+#include "third_party/tonic/logging/dart_invoke.h"
+#include "third_party/tonic/typed_data/dart_byte_data.h"
+
+namespace flutter {
+namespace {
+void FreeFinalizer(void* isolate_callback_data, void* peer) {
+  free(peer);
+}
+}  // namespace
+
+PlatformMessageResponseDartPort::PlatformMessageResponseDartPort(
+    Dart_Port send_port,
+    int64_t identifier,
+    const std::string& channel)
+    : send_port_(send_port), identifier_(identifier), channel_(channel) {
+  FML_DCHECK(send_port != ILLEGAL_PORT);
+}
+
+void PlatformMessageResponseDartPort::Complete(
+    std::unique_ptr<fml::Mapping> data) {
+  is_complete_ = true;
+  Dart_CObject response_identifier = {
+      .type = Dart_CObject_kInt64,
+  };
+  response_identifier.value.as_int64 = identifier_;
+  Dart_CObject response_data = {
+      .type = Dart_CObject_kExternalTypedData,
+  };
+  uint8_t* copy = static_cast<uint8_t*>(malloc(data->GetSize()));
+  memcpy(copy, data->GetMapping(), data->GetSize());
+  response_data.value.as_external_typed_data.type = Dart_TypedData_kUint8;
```

Comment:
Should this use `Dart_CObject.value.as_typed_data` instead of external typed data?

`dart_native_api.h` says that Dart will handle copying the `as_typed_data` buffer into a Dart object

Reply:
You mean performance-wise?  I believe the logic we have elsewhere is that if it is less than 1KB use typed data, otherwise we use external typed data.

Reply:
The code above is making a copy of the data in order to construct an external typed data.

So I don't think there is any advantage to using external typed data here.  This could instead create a standard typed data where Dart would manage the allocation and do the copy.

Reply:
I looked into doing this but typedata would require a const-cast because `response_data.value.as_typed_data.values = data->GetMapping();` doesn't work.  I'll leave this as is and file a bug for dart and add it as a comment.  `as_typed_data.values` should be const if it is doing a copy.

Reply:
Done. Added comment and linked Dart issue: https://github.com/dart-lang/sdk/issues/49827

Reply:
I think that doing the const cast here would be reasonable.  The Dart API is a C API that does not closely follow C++-style conventions for usage of const.

Reply:
It's not really c++ style conventions, it's c convention too.  Look at the signature of memcpy, the source parameter is `const`.  There's no good reason to have a non-const pointer.  If they have a bad reason it should be non-const then a const-cast isn't safe.  They might fix it before this lands anyway.